// ice_security/ice_security.js
'use strict';

var _ = require('underscore');
var kOK = 0;
var kUNAUTH = 1;
var kNOROLE = 2;

module.exports.kOK = kOK;
module.exports.kUNAUTH = kUNAUTH;
module.exports.kNOROLE = kNOROLE;

var role_type = function(req, prole){
  if (req.isAuthenticated()){
    if(prole===undefined || prole===null) return kOK;

    var user_role = '';
    if(req.user && req.user.role){
      user_role = req.user.role;
    }

    if(Object.prototype.toString.call( prole ) === '[object Array]')
    {
      var res = _.some(prole, function(role){
        role = role || '';
        return role=='any' || user_role.indexOf(role)>=0;
      });
      return (res) ? kOK : kNOROLE;
    }else
    {
      console.log(user_role);
      prole = prole || '';
      var res2 =  prole=='any' || user_role.indexOf(prole)>=0;
      return (res2) ? kOK : kNOROLE;
    }
  }
  return kUNAUTH;
};

module.exports.role_type = role_type;

module.exports.role_is = function(req, prole){
  var sec_res = role_type(req, prole);
  if(sec_res === kOK) return true;
  return false;
};

module.exports.has_role_middleware = function(role){
  return function (req, res, next){
    var sec_res = role_type(req, role);
    if(sec_res === kOK) return next();
    return res.send(500, JSON.stringify({error:'bad auth', auth:sec_res}));
  };
};
